performance-tests/bench_2/test_controller/ScenarioManager.cpp

```diff
@@ -201,107 +202,254 @@ void ScenarioManager::customize_configs(std::map<std::string, std::string>& work
   }
 }
 
-AllocatedScenario ScenarioManager::allocate_scenario(
-  const ScenarioPrototype& scenario_prototype, Nodes& available_nodes, bool debug_alloc)
+bool ScenarioManager::is_matched(const std::string& str, const std::string& pat) const
+{
+  if (pat.empty() ||
+      (!OpenDDS::DCPS::is_wildcard(pat.c_str()) && pat == str) ||
+      (OpenDDS::DCPS::is_wildcard(pat.c_str()) && ACE::wild_match(str.c_str(), pat.c_str(), true, true))) {
+    return true;
+  }
+  return false;
+}
+
+AllocatedScenario ScenarioManager::allocate_scenario(const ScenarioPrototype& scenario_prototype,
+  const Nodes& available_nodes, bool debug_alloc)
 {
   if (scenario_prototype.nodes.length() == 0 && scenario_prototype.any_node.length() == 0) {
     throw std::runtime_error("Scenario Prototype is empty!");
   }
 
-  // Get Maximum and Exclusive Node Counts
-  unsigned maximum_nodes = 0;
-  unsigned exclusive_nodes = 0;
-  for (unsigned i = 0; i < scenario_prototype.nodes.length(); i++) {
-    if (scenario_prototype.nodes[i].exclusive) {
-      exclusive_nodes += scenario_prototype.nodes[i].count;
+  // Get a set of matched node controllers for each node prototype
+  typedef std::vector<NodeController::NodeId> MatchedNodeControllers;
+  std::vector<MatchedNodeControllers> matched_ncs(scenario_prototype.nodes.length());
+  for (unsigned i = 0; i < scenario_prototype.nodes.length(); ++i) {
+    const NodePrototype& nodeproto = scenario_prototype.nodes[i];
+    for (Nodes::const_iterator it = available_nodes.begin(); it != available_nodes.end(); ++it) {
+      if (is_matched(it->second.name.in(), nodeproto.name_wildcard.in())) {
+        matched_ncs[i].push_back(it->first);
+      }
     }
-    maximum_nodes += scenario_prototype.nodes[i].count;
-  }
-  // If there are enough nodes, each any node worker can get one to itself.
-  for (unsigned protoworker_i = 0; protoworker_i < scenario_prototype.any_node.length(); protoworker_i++) {
-    maximum_nodes += scenario_prototype.any_node[protoworker_i].count;
   }
 
-  // Get Minimum Node Count
-  unsigned minimum_nodes = 0;
-  minimum_nodes = exclusive_nodes;
-  if (maximum_nodes - exclusive_nodes) { // If there are any non-exclusive node configs
-    minimum_nodes++; // Then there just has to be at least one node for them
+  struct NcMetaData {
```

Comment:
Let's move this and the queue typedef into a new header.

Reply:
Should it be moved to a header? It just contains extra data needed to compute the allocation, and I don't think it'll be used somewhere else.

Reply:
In general, I prefer not to see local structs and this function's already pretty large. Perhaps you can pull some of this logic into a separate helper class (AllocationHelper, etc) that's easier to create a test against, and this struct can be part of that helper class?